src/server/pbsd_init.c

```diff
@@ -828,67 +746,24 @@ pbsd_init(int type)
 	/* get jobs from DB */
 	obj.pbs_db_obj_type = PBS_DB_JOB;
 	obj.pbs_db_un.pbs_db_job = &dbjob;
-	state = pbs_db_cursor_init(conn, &obj, NULL);
-	if (state == NULL) {
-		sprintf(log_buffer, "%s", (char *) conn->conn_db_err);
+	rc = pbs_db_search(conn, &obj, NULL, (query_cb_t)&recov_job_cb);
+	if (rc == -1) {
+		pbs_db_get_errmsg(PBS_DB_ERR, &conn_db_err);
+		sprintf(log_buffer, "%s", conn_db_err);
```

Comment:
PBS_DB_ERR is handled within pbs_db_get_errmsg. If strdup itself fails, then PBS will be running out of memory.

Thinking on pbs_db_get_errmsg(). Why does it have to strdup all the time and free in the caller. Have you thought of a buffer internal to libdb and reusing that?

Reply:
For me, this looks simple. You call pbs_db_get_errmsg only when a libdb API fails. And caller knows it failed. Maintaining cache inside libdb adds the overhead of managing that cache which I dont think is a good idea.

Reply:
On the contrary, libdb is delegating its responsibility to check the buffer allocation failure to the caller. Caller has to check the buffer is valid and free after use. Which can be eliminated with a static buffer inside libdb.

Reply:
I still feel the caller should take care of this since it knows when it is interested in getting the error message and this way we can keep libdb simple.

Reply:
I do not understand why it has to be complex? Maybe something we can take up later!